docs/opcodes_details.md

```diff
@@ -0,0 +1,340 @@
+
+Based on `src/program_generator/data/opcodes.csv` and `src/program_generator/data/selection.csv`.
+
+## Takeaways
+
+See below for a detailed OPCODE-by-OPCODE description of parameters and concerns.
+This is a proposed list of things to tackle, in order from important to negligible:
+
+- parametrize values on the stack taken for stack-only operations, see sections `0x01,ADD`, `0x0a,EXP`
+- parametrize amount of memory allocation using memory read/write OPCODEs, see section `0x52,MSTORE`
+- parametrize `JUMP` AND `JUMPI` calls circumstances, see their sections
+- parametrize `JUMPDEST` calls circumstances, see section
+- parametrize amount of memory copying and padding, see sections `0x37,CALLDATACOPY`, `0x39,CODECOPY`, `0x3e,RETURNDATACOPY`
+- parametrize size of the stack, see sections `0x50,POP` `0x60,PUSH1` `0x80,DUP1`
+- parametrize amount of memory allocation using OPCODEs other than MLOAD, see sections `0x37,CALLDATACOPY`, `0x39,CODECOPY`, `0x3e,RETURNDATACOPY`, `0xf3,RETURN`
+- parametrize context values, see sections `0x30,ADDRESS` `0x32,ORIGIN` `0x33,CALLER` `0x34,CALLVALUE` `0x3a,GASPRICE` `0x41,COINBASE`
+  - in particular parametrize BigInt-based variables: `0x3a,GASPRICE`, `0x34,CALLVALUE`
+
+## `0x00,STOP`
+
+`0x00,STOP,0,zero,0,0,Halts execution.,`
+
+- **geth**: does completely nothing
+- **evmone**: sets execution state to success and does nothing
+- **openethereum**: sets execution state to done and does nothing
+
+can be dropped
+
+## `0x01,ADD` and other binary/ternary arithmetic/logical OPCODEs
+
+The "simple" ones
+
+```
+0x01,ADD,3,verylow,2,1,Addition operation,
+0x02,MUL,5,low,2,1,Multiplication operation.,
+0x03,SUB,3,verylow,2,1,Subtraction operation.,
+0x04,DIV,5,low,2,1,Integer division operation.,
+0x05,SDIV,5,low,2,1,Signed integer division operation (truncated).,
+0x06,MOD,5,low,2,1,Modulo remainder operation,
+0x07,SMOD,5,low,2,1,Signed modulo remainder operation.,
+0x08,ADDMOD,8,mid,3,1,Modulo addition operation.,
+0x09,MULMOD,8,mid,3,1,Modulo multiplication operation.,
+0x0b,SIGNEXTEND,5,low,2,1,Extend length of two’s complement signed integer.,
+0x10,LT,3,verylow,2,1,Less-than comparison.,
+0x11,GT,3,verylow,2,1,Greater-than comparison.,
+0x12,SLT,3,verylow,2,1,Signed less-than comparison.,
+0x13,SGT,3,verylow,2,1,Signed greater-than comparison.,
+0x14,EQ,3,verylow,2,1,Equality comparison.,
+0x15,ISZERO,3,verylow,1,1,Simple not operator.,
+0x16,AND,3,verylow,2,1,Bitwise AND operation.,
+0x17,OR,3,verylow,2,1,Bitwise OR operation,
+0x18,XOR,3,verylow,2,1,Bitwise XOR operation.,
+0x19,NOT,3,verylow,1,1,Bitwise NOT operation.,
+0x1a,BYTE,3,verylow,2,1,Retrieve single byte from word,
+0x1b,SHL,3,verylow,2,1,Left shift operation,
+0x1c,SHR,3,verylow,2,1,Logical right shift operation,
+0x1d,SAR,3,verylow,2,1,Arithmetic (signed) right shift operation,
+```
+
+They pop and push, or pop and peek and then modify stack in-place.
+
+### Parameters
+
+2 or 3 values on the stack, should not have impact on cost.
```

Comment:
Many of these instruction are actually sensitive to the input arguments. One reason is that argument values may be out of the core instruction operational range in such case the output being some arbitrary values (e.g. left shift by large value produces 0 but this case may be handled separately in the implementation). Other reason is implementation complexity which may have multiple control-flow paths to follow (e.g. my division implementation actually has 3 different algorithms depending on the divisor size).